### PR TITLE
New module: support Alibaba Cloud ECS (cloud/alicloud/alicloud_instance)

### PR DESCRIPTION
**Pull request by xiaozhu36**
This is reopen PR of #25975 

SUMMARY

Added a new module alicloud_instance.py that can manage alicloud ecs instance.

COMPONENT NAME

lib/ansible/module_utils/alicloud_ecs.py
lib/ansible/modules/cloud/alicloud/alicloud_instance.py
lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py

ANSIBLE VERSION

2.4.0

### CODE REVIEW

**Comment by dharmabumstead** [comment]
Edit pending. Please don't merge.

**Review by dharmabumstead — DISMISSED** [review]
The guide_alicloud file will need to be moved into the scenario_guides directory.

**Comment by xiaozhu36** [comment]
Hi @dharmabumstead I have moved guide_alicloud file into the scenario_guides directory, and please check it. Thanks a lot.

**Comment by xiaozhu36** [comment]
HI @dharmabumstead @gundalow Is there any progress about this PR? Do I need to do more works?

File: docs/docsite/rst/scenario_guides/guide_alicloud.rst [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+For storing these in a vars_file, ideally encrypted with `ansible-vault <https://docs.ansible.com/ansible/2.4/vault.html>`_ considering its security::
+
+    ---
+    alicloud_access_key: "--REMOVED--"
+    alicloud_secret_key: "--REMOVED--"
+
+Note that if you store your credentials in vars_file, you need to refer to them in each Alicloud-module. For example::
+
+    - alicloud_instance:
+      alicloud_access_key: "{{alicloud_access_key}}"
```
Comment (ryansb) [inline docs/docsite/rst/scenario_guides/guide_alicloud.rst]: Incorrect indentation - these lines need 2 more spaces.

File: docs/docsite/rst/scenario_guides/guide_alicloud.rst [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+There are a number of modules to create ECS instance, disk, VPC, VSwitch, Security Group and other resources.
+
+An example of making sure there are only 5 instances tagged 'NewECS' and other resources as Alicloud Module follows.
+
+In the example below, the ``count`` of instances is set to 5. This means if there are 0 instances already existing, then
+5 new instances would be created. If there were 2 instances, only 3 would be created, and if there were 8 instances,
+3 instances would be terminated.
+
+If ``count_tag`` is not specified, ``coun`` would use ``instance_name`` to create specified number of instances.
```
Comment (ryansb) [inline docs/docsite/rst/scenario_guides/guide_alicloud.rst]: s/coun/count/

File: lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        module.fail_json(msg='footmark required for the module alicloud_instance_facts')
+
+    ecs = ecs_connect(module)
+
+    instances = []
+    instance_ids = []
+    ids = module.params['instance_ids']
+    names = module.params['instance_names']
+    zone_id = module.params['alicloud_zone']
+    if ids and (not isinstance(ids, list) or len(ids)) < 1:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py]: This conditional seems incorrect - parenthesis are closed too early. 

I think it should read `if ids and (not isinstance(ids, list) or len(ids) < 1):`

File: lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        module.fail_json(msg='instance_ids should be a list of instances, aborting')
+
+    if names and (not isinstance(names, list) or len(names)) < 1:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py]: Same here

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        - Internet charge type of ECS instance.
+      default: 'PayByBandwidth'
+      choices: ['PayByBandwidth', 'PayByTraffic']
+    max_bandwidth_in:
+      description:
+        - Maximum incoming bandwidth from the public network, measured in Mbps (Mega bit per second).
+      default: 200
+    max_bandwidth_out:
+      description:
+        - Maximum outgoing bandwidth to the public network, measured in Mbps (Mega bit per second).
```
Comment (ryansb) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: "Megabits per second"

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+      default: 0
+    host_name:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: Please add an alias for "hostname"

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+      description:
+        - Instance host name.
+    password:
+      description:
+        - The password to login instance. After rebooting instances, the modified password would be take effect.
```
Comment (ryansb) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: `modified password will take effect`

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+      default: False
+      type: bool
+    instance_tags:
+      description:
+        - A hash/dictionaries of instance tags, to add to the new instance or for starting/stopping instance by tag. C({"key":"value"})
+      aliases: ["tags"]
+    sg_action:
+      description:
+        - The action of operating security group.
+      choices: ['join', 'leave']
```
Comment (ryansb) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: Adding/removing security groups is usually implemented like this:

```
groups:
- {identifier of security group}
- {identifier of second security group}
purge_groups: true/false
```
That way if a user specifies a set of groups, the instance will be added if they are not already on it. If they want to leave groups, they specify the ones that must remail and then set `purge_groups: true`. For an example, see the `ec2_group` module.

**Review by ryansb — DISMISSED** [review]

**Review by acozine — CHANGES_REQUESTED** [review]
Thanks for submitting a guide for the Alibaba Cloud modules. I reviewed the Scenario Guide and pushed some changes to https://github.com/xiaozhu36/ansible/pull/1.

**Review by acozine — APPROVED** [review]

**Comment by xiaozhu36** [comment]
HI @ryansb @acozine Thanks for your review and I have improved the pr according to your suggestion. Please check them.

Regards,
He Guimin

**Comment by xiaozhu36** [comment]
HI @ryansb @acatton @gundalow This PR has opened more than one year and the module version has modified from 2.5 to 2.6. Now, the CI error show me 
```
ERROR: lib/ansible/modules/cloud/alicloud/ali_instance.py:0:0: E307 version_added should be 2.8. Currently 2.7 (75%)
```
Please let me know when it can be merged?

**Review by ryansb — COMMENTED** [review]
Needs rerun in CI and update for version_added to 2.7

**Comment by xiaozhu36** [comment]
HI @ryansb It seems like `2.7` does not work. What I should do?

File: lib/ansible/modules/cloud/alicloud/ali_instance_facts.py [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: ali_instance_facts
+version_added: "2.7"
```
Comment (acozine) [inline lib/ansible/modules/cloud/alicloud/ali_instance_facts.py]: @xiaozhu36 you should update this version number from 2.7 to 2.8.

We have already released 2.7, so this module will be added as of version 2.8.

File: docs/docsite/rst/dev_guide/developing_module_utilities.rst [hunk 10]
```diff
@@ -1,18 +1,16 @@
...
-  from ansible.module_utils.basic import AnsibleModule
+  from ansible.module_utils.basic import *
 
 
-The following is a list of ``module_utils`` files and a general description. The module utility source code lives in the ``./lib/ansible/module_utils`` directory under your main Ansible path - for more details on any specific module utility, please see the source code.
-
-.. include:: shared_snippets/licensing.txt
+The following is a list of module_utils files and a general description. The module utility source code lives in the `./lib/module_utils` directory under your main Ansible path - for more details on any specific module utility, please see the source code.
 
+- alicloud_ecs.py - Definitions and utilities for modules working with Alibaba Cloud ECS.
```
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_module_utilities.rst:16]: This new line should be the only change to the `docs/docsite/rst/dev_guide/developing_module_utilities` file. Please remove/undo the other changes.

File: lib/ansible/modules/cloud/alicloud/ali_instance_facts.py (hunk 9 above)
Comment (xiaozhu36) [reply lib/ansible/modules/cloud/alicloud/ali_instance_facts.py]: Done. When release the version 2.8?

File: docs/docsite/rst/dev_guide/developing_module_utilities.rst (hunk 10 above)
Comment (xiaozhu36) [reply docs/docsite/rst/dev_guide/developing_module_utilities.rst:16]: Done.

**Review by ryansb — APPROVED** [review]

**Pull request merged**